include/dali/core/per_stream_pool.h

```diff
@@ -0,0 +1,207 @@
+ *  - user-provided functor to create new objects, if none are available?
+ *
+ * @tparam T                              type of the managed object.
+ * @tparam mutex_type                     type of the synchronization object used to protect
+ *                                        internal structures
+ * @tparam reuse_pending_on_same_stream   if true, subsequent requests for an object
+ *                                        specifying same stream may return an object associated
+ *                                        with an unfinished job on that stream
+ */
+template <typename T, typename mutex_type = std::mutex, bool reuse_pending_on_same_stream = true>
+class PerStreamPool {
+ private:
+  struct ListNode;
+  using ListNodeUPtr = std::unique_ptr<ListNode>;
+  struct ListNode {
+    T object;
+    int device_id = -1;
+    cudaStream_t stream = nullptr;
+    CUDAEvent event;
+    ListNodeUPtr next;
+
+    template <typename... ObjectConstructorArgs>
+    ListNode(int device_id, cudaStream_t stream, ObjectConstructorArgs&&... args)
+    : object(std::forward<ObjectConstructorArgs>(args)...)
+    , device_id(device_id)
+    , stream(stream)
+    , event(CUDAEvent::Create()) {
+    }
+
+    ~ListNode() {
+      event.reset();
+      DeleteNonRecursive(std::move(next));
+    }
+
+    static void DeleteNonRecursive(ListNodeUPtr ptr) {
+      while (ptr) {
+        auto tmp = std::move(ptr->next);
+        // ptr will be deleted, but its next pointer is null now, so we'll not recurse
+        ptr = std::move(tmp);
+      }
+    }
+  };
+
+ public:
+  class ObjectLease {
+   public:
+    ObjectLease(ObjectLease &&) = default;
+    ~ObjectLease() { owner->Release(std::move(node)); }
+    operator T*() const { return &node->object; }
+    T &operator*() const { return node->object; }
+    T *operator->() const { return &node->object; }
+    explicit operator bool() const noexcept { return node != nullptr; }
+   private:
+    ObjectLease(PerStreamPool *owner, ListNodeUPtr node) : owner(owner), node(std::move(node)) {}
+    PerStreamPool *owner;
+    ListNodeUPtr node;
+    friend class PerStreamPool;
+  };
+
+  ObjectLease Get(int device_id, cudaStream_t stream) {
+    std::lock_guard<mutex_type> guard(lock_);
+    RecycleFinished();
+    if (reuse_pending_on_same_stream) {
+      if (auto ret = GetPending(stream))
+          return { this, std::move(ret) };
+    }
+    return { this, GetFromDevicePool(device_id, stream) };
+  }
+
+  ObjectLease Get(cudaStream_t stream) {
+    int device = -1;
+    cudaGetDevice(&device);
+    return Get(device, stream);
+  }
+
+ private:
+  friend class ObjectLease;
+
+  void Release(std::unique_ptr<ListNode> node) {
+    cudaEventRecord(node->event, node->stream);
+    std::lock_guard<mutex_type> guard(lock_);
+    node->next = std::move(pending_);
+    pending_ = std::move(node);
+  }
+
+  ListNodeUPtr GetPending(cudaStream_t stream) {
+    for (ListNodeUPtr *pptr = &pending_; *pptr; pptr = &(*pptr)->next) {
```

Comment:
Why pointer to ListNodeUPtr ?

Reply:
We need a double pointer to modify it (to reconnect the list).